### PR TITLE
Add fallbacks

### LINKED ISSUE #38: Add fallback to `calculate_step`
Consider adding a fallback call for `type_of_method = 2` and `nlls_method = 3|4`
if `call regularization_solver` fails, then don't terminate with error but try `call solve_galahad`.

https://github.com/ralna/RALFit/blob/b9f438c25874379a71922f7c18f352722d7cf55c/libRALFit/src/ral_nlls_internal.f90#L1192-L1207

The issue is raised when for some reason Delta is such that `Ashift = A + 1/Delta` does not guarantee `Ashift` is SPD and the call to `SPD_solve` fails.

### PR DESCRIPTION
**Pull request by tyronerees**
Fixes #38 

The proposed solution tries the method picked in options, and then cycles through until one of the method calculates a step.  The method is reset to that chosen by the user at each iteration.

The changes apply the same method to the trust region part of the code, too, as it could also be the case that one of the four methods there fails to calculate a step.

### CODE REVIEW

**Comment by tyronerees** [comment]
I still need to get the tests passing (options that would have failed now pass in the c test deck), and I also want to see if I can increase the coverage, as this made a significant drop.

**Comment by tyronerees** [comment]
Fixed the deallocation logic by making workspaces `intent(inout)` -- if the review is ok, this should be ready to merge to master now.

**Comment by tyronerees** [comment]
Commit a0fddef also fixes #50 --> I couldn't reproduce it from the master branch (although I believe it's an issue there too), so adding it to this pull request.  

The issue was that the logic tests `inform%iter` to see if the banner should be printed, but this is only updated at the end of each successful iteration: `w%iter` holds the true iteration count.

File: libRALFit/src/ral_nlls_internal.f90 [hunk 1]
```diff
@@ -1145,11 +1145,15 @@ RECURSIVE SUBROUTINE calculate_step(J,f,hf,g,X,md,md_gn,n,m,use_second_derivativ
...
-             ! only use fallback if the options say it's ok
-             if (.not. options%allow_fallback_method) Go To 100
+             ! If we're using a method for the second time, or
+             ! we're not allowing a fallback, then bail...
+             if ( (subproblem_method == options%nlls_method) .or. & 
+                  (.not. options%allow_fallback_method)) Go To 100
+
+             if (buildmsg(5,.False.,options)) then
+                Write(rec(1), Fmt=3040)
+                Call printmsg(3,.False.,options,1,rec)
```
Comment (talassio) [inline libRALFit/src/ral_nlls_internal.f90]: @tyronerees: Should match with buildmsg level 5.
`Call printmsg(5,.False.,options,1,rec)`

**Pull request merged**